File: src/IO/validator.py
import re

class Validator:
    lines: int
    columns: int
    binary_matrix: list[list[int]]

    def __init__(self, instance_path: str):
        with open(instance_path) as f:
            self.lines, self.columns = map(int, f.readline().split())
            self.binary_matrix = [[0 for _ in range(self.columns)] for _ in range(self.lines)]

            for i in range(self.lines):
                values = list(map(int, f.readline().split()))
                edges = values[1:]
                for e in edges:
                    self.binary_matrix[i][e - 1] = 1

    def validate(self, perm: list[int], cost: int):
        v1 = self.contains_all_columns(perm)
        v2 = self.count_one_blocks(perm)
        if not (v1 and (cost == v2)):
            raise Exception(f"Invalid solution: {v1} or {v2} != {cost}")
        return True

    def count_one_blocks(self, perm: list[int]) -> int:
        perm_matrix = [[row[j] for j in perm] for row in self.binary_matrix]
        bitstring = "-".join("".join(str(cell) for cell in row) for row in perm_matrix)
        blocks = re.findall(r'1+', bitstring)
        return len(blocks)

    def contains_all_columns(self, perm: list[int]) -> bool:
        return set(perm) == set(range(self.columns))

File: src/CBMLKH/Validator.cpp
#include "Validator.hpp"

#include <fstream>
#include <stdexcept>
#include <unordered_set>

Validator::Validator(const std::string& instancePath) {
    std::ifstream f(instancePath);
    if (!f.is_open()) throw std::runtime_error("Error opening file: " + instancePath);

    f >> lines >> columns;
    binaryMatrix.assign(lines, std::vector<int>(columns, 0));

    for (int i = 0; i < lines; i++) {
        int count;
        f >> count;
        for (int j = 0; j < count; j++) {
            int e;
            f >> e;
            binaryMatrix[i][e - 1] = 1;
        }
    }
}

bool Validator::validate(const std::vector<int>& perm, int cost) {
    bool v1 = containsAllColumns(perm);
    int v2 = countOneBlocks(perm);
    if (!v1 || cost != v2)
        throw std::runtime_error("Invalid solution: containsAll=" + std::to_string(v1) + " countBlocks=" + std::to_string(v2) +
                                 " != cost=" + std::to_string(cost));
    return true;
}

int Validator::countOneBlocks(const std::vector<int>& perm) {
    int blocks = 0;
    for (int row = 0; row < lines; row++) {
        bool inBlock = false;
        for (int col : perm) {
            if (binaryMatrix[row][col] == 1) {
                if (!inBlock) {
                    blocks++;
                    inBlock = true;
                }
            } else {
                inBlock = false;
            }
        }
    }
    return blocks;
}

bool Validator::containsAllColumns(const std::vector<int>& perm) {
    std::unordered_set<int> permSet(perm.begin(), perm.end());
    for (int i = 0; i < columns; i++)
        if (permSet.find(i) == permSet.end()) return false;
    return true;
}